src/core/expr/head_reduce_unary.cc

```diff
@@ -1573,7 +1618,8 @@ Workframe Head_Reduce_Unary::evaluate_n(
       case Op::MIN:
       case Op::MAX:
       case Op::FIRST:
-      case Op::LAST:   fn = compute_gfirstlast; break;
+      case Op::LAST:
+      case Op::COUNTNA:fn = compute_gfirstlast; break;
```

Comment:
I don't think this will work. We need a test for how `countna()` behaves when computed within group-by.

Reply:
I agree. I didn't look properly into this. Can you give an example for the expected behaviour of `countna()` in the presence of grouby and contrast that output with how `count()` would behave in the presence of groupby, thanks.

Reply:
`count(f.X)` returns the number of valid (non-NA) elements in column X in each group. The `countna(f.X)` is supposed to return the number of invalid (NA) elements in column X in each group. The `count()` function without an argument returns the total number of elements in each group. Thus, the following invariant must hold:
```
countna(f.X) == count() - count(f.X)
```

Reply:
Let me make sure we are on the same page here. The following frame
```python
df = dt.Frame({"group1": ["A", "A", None, "B", "B", "A"],
               "group2": [1, 0, 1, 1, 1, 1],
               "var": [343, 345, 567, 345, 2, 212]})

print(df[:, dt.count()])
print(df[:, dt.count(f[:])])

print(df[:, dt.countna()])
print(df[:, dt.countna(f[:])])

```
gives

```python
   | count
   | int32
-- + -----          # df[:, dt.count()]
 0 |     6
[1 row x 1 column]

   | group1
   |  int64
-- + ------          # df[:, dt.count(f.group1)]       
 0 |      5
[1 row x 1 column]

   |   C0
   | void
-- + ----          # df[:, dt.countna()]
 0 |   NA
[1 row x 1 column]

   | group1
   |  int64
-- + ------          # df[:, dt.countna(f.group1)]
 0 |      1
[1 row x 1 column]

```

The relationship `countna(f.X) == count() - count(f.X)` does hold. The only weird result here is for `dt.countna()`, which takes the `else` path in head_reduce_unary.cc:1614 and it needs fixing. is that correct?

Reply:
Yeah, I don't think `dt.countna()` without an argument makes sense. Should throw a type error immediately.

But my original concern was about `countna()` in the presence of a groupby clause, i.e. the following two expressions should produce same results:
```
DT[:, [dt.countna(f.var), dt.count() - dt.count(f.var)], by(f.group1)]
``` 
(throw some NAs into the column `var` so that the example is not too trivial).

Reply:
This is what I see now
```python
df = dt.Frame({"group1": ["A", "A", None, "B", "B", "A"],
                "var": [343, None, 567, None, 2, 212]})
                                                           
print(df[:, [dt.countna(f.var), dt.count() - dt.count(f.var)], dt.by(f.group1)])
```
gives
```python
  | group1    var     C0
   | str32   int64  int64
-- + ------  -----  -----
 0 | NA          0      0
 1 | A           1      1
 2 | B           1      1
[3 rows x 3 columns]

```
isn't this the expected output? 

Perhaps you can help me understand this. One thing I am not sure is under what circumstances the program takes the `else` path in `head_reduce_unary.cc:1614`? Because that was your initial concern in the review and I am not sure when that happens.

Reply:
Hmm, could it be group function applied to another group function?

Reply:
not sure, perhaps an example to illustrate what you mean?

Reply:
Suppose you have `DT = dt.Frame(G=[1,1,1,2,2,2], V=[None, None, None, None, 3, 5])`, then simple aggregate produces
```
>>> DT[:, [f.V, dt.mean(f.V)], by(f.G)]
   |     G      V      V.0
   | int32  int32  float64
-- + -----  -----  -------
 0 |     1     NA       NA
 1 |     1     NA       NA
 2 |     1     NA       NA
 3 |     2     NA        4
 4 |     2      3        4
 5 |     2      5        4
[6 rows x 3 columns]
```
Now imagine we applied `countna()` to both of these columns:
```
>>> DT[:, [dt.countna(f.V), dt.countna(dt.mean(f.V))], by(f.G)]
   |     G      V    V.0
   | int32  int64  int64
-- + -----  -----  -----
 0 |     1      3      3
 1 |     2      1      0
[2 rows x 3 columns]
```
is this the result you're getting?

Reply:
I see

```python
  |     G      V      V.0
   | int32  int64  float64
-- + -----  -----  -------
 0 |     1      3       NA
 1 |     2      1        4
[2 rows x 3 columns]

```
and `countna` on the third column also take the `else` path in `head_reduce_unary.cc:1614` That is a clear example where it fails.